File: python/problems/leetcode/0155_min_stack/solution.py
"""LeetCode 155: Min Stack."""

from __future__ import annotations


class MinStack:
    def __init__(self) -> None:
        self._st: list[int] = []
        self._mn: list[int] = []

    def push(self, val: int) -> None:
        self._st.append(val)
        if not self._mn or val <= self._mn[-1]:
            self._mn.append(val)
        else:
            self._mn.append(self._mn[-1])

    def pop(self) -> None:
        self._st.pop()
        self._mn.pop()

    def top(self) -> int:
        return self._st[-1]

    def getMin(self) -> int:
        return self._mn[-1]


if __name__ == "__main__":
    s = MinStack()
    s.push(-2)
    s.push(0)
    s.push(-3)
    assert s.getMin() == -3
    s.pop()
    assert s.top() == 0
    assert s.getMin() == -2
    print("0155 OK")

File: cpp/problems/leetcode/0155_min_stack/solution.cpp
// LeetCode 155
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class MinStack {
public:
    void push(int val) {
        st.push_back(val);
        mn.push_back(mn.empty() ? val : min(mn.back(), val));
    }
    void pop() {
        st.pop_back();
        mn.pop_back();
    }
    int top() { return st.back(); }
    int getMin() { return mn.back(); }

private:
    vector<int> st, mn;
};

int main() {
    MinStack s;
    s.push(-2);
    s.push(0);
    s.push(-3);
    assert(s.getMin() == -3);
    s.pop();
    assert(s.top() == 0);
    assert(s.getMin() == -2);
    cout << "0155 OK" << endl;
    return 0;
}
